src/libos/crates/block-device/src/block_id.rs

```diff
@@ -0,0 +1,84 @@
+//! Block id.
+use core::convert::TryInto;
+use core::ops::{Add, AddAssign, Sub, SubAssign};
+
+use crate::prelude::*;
+use errno::return_errno;
+
+type Id = usize;
+type Offset = usize;
+
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BlockId(Id);
+
+impl BlockId {
+    pub const fn new(id: Id) -> Self {
+        debug_assert!(
+            BLOCK_SIZE.saturating_mul(id) <= usize::MAX as usize,
+            "block id is too big"
+        );
+        Self(id)
+    }
+
+    pub fn id(&self) -> Id {
+        self.0
+    }
+
+    pub fn to_offset(&self) -> Offset {
+        self.0 * BLOCK_SIZE
+    }
+
+    pub fn from_offset(offset: Offset) -> Result<Self> {
```

Comment:
I am wondering whether this method should be fallible or not. I check all modifications: there are many manual conversion from a byte offset to a block ID.

```rust
let block_id = BlockId::new(offset / BLOCK_SIZE);
```

But no usage of `from_offset` is found at all. This result seems to suggest that it is quite common to convert an _unaligned_ offset to a block ID and this is considered harmless.

So I would suggest having two methods to convert an offset to a block ID, one is infallible, the other is fallible.

```rust
pub fn from_byte_offset(offset: usize) -> Self {
    Self(offset / BLOCK_SIZE)
}

pub fn from_byte_offset_aligned(offset: usize) -> Result<Self> {
    Self::check_align(offset)?;
    Ok(Self((offset / BLOCK_SIZE) as _))
}
```

And `from_byte_offset` should be the one that is most often used.

Reply:
So do we still need a `new()`(or rename to `from_raw()`) to take `raw_bid: RawBId` as input, or just use two infallible/fallible fns that only takes `offset: usize`

Reply:
We need to keep `new`, which should be viewed as the default constructor for `BlockId`.